### DISCUSSION TITLE
Melhor forma de trabalhar com retorno de API REST

### DISCUSSION THREAD

**Discussion by seadramdesenv** [Flutterando/forum#21] [Discussões gerais]
Boa tarde pessoal,
Uma coisa que não confio que faço da melhor forma é o tratamento de retorno de API REST, pois temos diversos status HTTP no response.
E me parece um pouco demais fazer uma tratativa para cada uma das possíveis respostas.
Nesse caso estou buscando uma forma de fazer uma tratativa mais abrangente porém não tão complexa como seria por cada status possível.

Se alguém tiver um exemplo de validação do retorno de API REST e puder compartilhar, creio que ajudaria bastante.

**Comment by JoaoWitorFelipe** [comment]
Não tenho exemplo mas tenho uma forma, pode fazer é tratar o primeiro número do status, por exemplo 2xx (independente se for 201 ou 200), sabemos que é sucesso do servidor, então dá para validar dessa forma. Se tiver autenticação/autorização como 401 ou 403, pode usar interceptors nos clients http para executar ações padrões como por exemplo redirecionar o usuário para uma tela de autenticação, se for 5xx exibir uma tela de erro. Enfim é isso, espero ter ajudado :)

**Reply by JoaoWitorFelipe** [reply]
Complementando (fugindo um pouco da ideia da pergunta), usar padrões como o Tripple por exemplo (da Flutterando) para evitar boilerplate de try catches e validando os 3 possíveis estados que são os mais comuns, os dados que quer mostrar, o erro e o loading também ajuda nas tratativas e padronização

**Comment by CriandoGames** [comment]
faria uma classe de tratativa com os erros listados pelo backend por exemplo 

erro 401, 500, 400.

switch(erro)
case : 400 {
 return a tratativa .. 
}
default : 
return resposta correta

**Reply by JoaoWitorFelipe** [reply]
Verdade, um enum ou classe com todos os códigos HTTP que o servidor usa é bem legal mesmo!

**Comment by sergiotucano** [comment]
Bom, uma boa forma é trabalhar com uma classe genérica para fazer as requisições e tratamentos, eu faço dessa forma : 

```
class AdapterHttp implements HttpClient {
AdapterHttp();

  @override
  Future<http.Response> request({@required Uri url, @required String method, Map body, Map headers, int timeout}) async {
     /// Implementação das requisições verbos get / post etc
   }


@override
  Future<void> handeResponse(final http.Response response, final String screen) async {

    /// Implementação da resposta da requisição

    var errorMessage = '';
    switch (response.statusCode) {
      case 200:       
      case 201:
      case 204:   
        return '';          
      case 400:             
        errorMessage = 'Requisição Inválida';       
        break;
      case 401:      
        errorMessage = 'Não Autorizado'; 
        break;
      case 403:     
        errorMessage = 'Proibido';      
        break;
      case 404:  
        errorMessage = 'Não Encontrado';     
        break;    
      case 422:       
        errorMessage = 'Entidade Não Processada';      
        break;             
      case 429:       
        errorMessage = 'Muitas Tentativas'; 
        break;
      case 503:
        errorMessage = 'Serviço Indisponível';
        break;
      default:
        errorMessage = 'Erro Interno Desconhecido';   
        break; 
    }

}
```

Assim funciona p qualquer requisição ou resposta , simples e funcional.
